src/build_helper/cc.rs

```diff
@@ -0,0 +1,496 @@
+        self.arch() == "i686"
+    }
+
+    pub fn is_x86_64(&self) -> bool {
+        self.arch() == "x86_64"
+    }
+
+    pub fn is_aarch64(&self) -> bool {
+        self.arch() == "aarch64"
+    }
+
+    pub fn is_windows(&self) -> bool {
+        self.os() == "windows"
+    }
+
+    pub fn is_mingw(&self) -> bool {
+        self.is_windows() && self.abi() == Some("gnu")
+    }
+
+    pub fn is_msvc(&self) -> bool {
+        self.abi() == Some("msvc")
+    }
+
+    pub fn is_linux(&self) -> bool {
+        self.os() == "linux"
+    }
+
+    pub fn is_ios(&self) -> bool {
+        self.os() == "ios"
+    }
+
+    pub fn is_android(&self) -> bool {
+        self.os() == "android"
+    }
+
+    pub fn is_darwin(&self) -> bool {
+        self.os() == "darwin"
+    }
+
+    /// Append the extension for the executables in this platform
+    pub fn with_exe_ext(&self, name : &str) -> String {
+        if self.is_windows() {
+            format!("{}.exe", name)
+        } else {
+            format!("{}", name)
+        }
+    }
+
+    /// Append the extension for the static libraries in this platform
+    pub fn with_lib_ext(&self, name : &str) -> String {
+        if self.is_msvc() {
+            format!("{}.lib", name)
+        } else {
+            format!("lib{}.a", name)
+        }
+    }
+
+    /// Get the file extension for the dynamic libraries in this platform.
+    pub fn dylib_ext(&self) -> &'static str {
+        if self.is_windows() {
+            "dll"
+        } else if self.is_darwin() {
+            "dylib"
+        } else {
+            "so"
+        }
+    }
+}
+
+impl fmt::Display for Triple {
+    fn fmt(&self, f : &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(&self.triple)
+    }
+}
+
+impl AsRef<OsStr> for Triple {
+    fn as_ref(&self) -> &OsStr {
+        &OsStr::new(&self.triple)
+    }
+}
+
+impl AsRef<Path> for Triple {
+    fn as_ref(&self) -> &Path {
+        &Path::new(&self.triple)
+    }
+}
+
+impl<'a> From<&'a Triple> for &'a str {
+    fn from(t : &'a Triple) -> &'a str {
+        &t.triple
+    }
+}
+
+impl<'a> From<&'a Triple> for String {
+    fn from(t : &'a Triple) -> String {
+        t.triple.clone()
+    }
+}
+
+/// Compile and assemble a static library. This will invoke the
+/// appropriate toolchain command (or LLVM tool) to compile the
+/// given list of files into object files and then invoke ar (or
+/// equivalent) to assemble them into a static library.
+///
+/// If a directory is given instead of a file, all supported files
+/// under that directory will be compiled.
+///
+/// Supported file types are:
+///
+/// `*.ll` --- LLVM byte code, will invoke `llc`
+///
+/// `*.S`, `*.c`, `*.cc`, `*.cpp` --- will invoke toolchain command
+/// defined by trait `Toolchain`.
+pub struct StaticLib {
+    toolchain : Box<Toolchain>,
+    llvm_tools : LLVMTools,
+    include_directories: Vec<PathBuf>,
+    files: Vec<PathBuf>,
+    src_dir : PathBuf,
+    build_dir : PathBuf,
+    with_llvm : bool
+}
```

Comment:
Much of this is almost a duplication around [`gcc-rs`](http://github.com/alexcrichton/gcc-rs). With Cargo we should be able to use out-of-tree dependencies, so leveraging what's already out there seems like a good idea?